### DISCUSSION TITLE
2.0 Ideas Discussion

### DISCUSSION THREAD

**Discussion by 0x19** [0x19/goesl#38] [Ideas]
Lets see what we can bring onto the table with 2.0 version.

**Comment by bcorrigeux** [comment]
the biggest thing i wanted to do but didnt end up building into the lib was a request/response handler so say you want to get a channel variable, currently you have to monitor the message channel and then map the uuid back to the request. it would be nice if the lib could handle this for the user. i have some code examples i can share but there might be a better way to do it inside the lib. this is mostly specific to an inbound esl connection, outbound is a little easier because its one call per connection

**Reply by 0x19** [reply]
If you can give example code, that would be awesome.

**Reply by bcorrigeux** [reply]
this is probably not the best way to do it inside of the library but this works on top of the library currently. im not sure we can get away from matching uuids because of the async nature of sending/receiving from freeswitch so this is how i went about it. I can do some digging again to see if i had a better plan than this when i was hoping to put it into the library but for now this is how im handling it
`var uuidMapLock sync.Mutex
var uuidMap = map[string]chan Message{}
var readMutex sync.Mutex

func requestUUIDMessages(uuid string, m map[string]interface{}) chan Message {
    //logDebug("requesting uuid: "+uuid, m)
    ch := make(chan Message, 2)
    uuidMapLock.Lock()
    uuidMap[uuid] = ch
    uuidMapLock.Unlock()
    return ch
}

func waitForUUIDMessages(ch chan Message, finishedTime time.Duration) (Message, error) {

    initialTime := 500 * time.Millisecond
    initialTicker := time.NewTicker(initialTime)
    defer initialTicker.Stop()
    completeTicker := time.NewTicker(finishedTime)
    defer completeTicker.Stop()

    for {
        select {
        case msg := <-ch:
            if typ, _ := msg.Headers["Content-Type"]; typ != "command/reply" {
                return msg, nil
            } else {
                //logDebug("received command reply, continuing", nil)
                initialTicker.Stop()
                continue
            }
        case <-initialTicker.C:
            logError("failed to receive start message back from freeswitch", nil, nil)
            return Message{}, errors.New("failed to receive ack from freeswitch")
        case <-completeTicker.C:
            return Message{}, nil
        }
    }
}`

**Comment by bcorrigeux** [comment]
another thing that i had to add was reconnect logic. currently in the code if there is an error, the lib does an os.Exit() which a library should never do. again i have some code built outside of the lib for reconnection that would probably be better suited to be inside the library

**Reply by 0x19** [reply]
Yeah. That was probably debugging that I've done at that time and left it in. Jezz.

**Reply by bcorrigeux** [reply]
i added a channel for text/disconnect messages in the plain-text / json / what whatever other types of messages so i could block waiting for a disconnect and reconnect if i got one. this is mainly useful for handling if freeswitch gets restarted but its possible to get disconnected for any reason so i think it should be  handled gracefully
